Get all the templates

wmde-manicki opened the pull request:
Task: https://phabricator.wikimedia.org/T123520
Demo: https://tools.wmflabs.org/file-reuse-test/api-continuation/

The problem that has been discovered using: `https://commons.wikimedia.org/wiki/File:Mona_Lisa,_by_Leonardo_da_Vinci,_from_C2RMF_retouched.jpg`. The page of that file has some 147 templates or so, while the tool has been only fetching 100 templates. That was usually OK but as the licence templates of the Mona Lisa didn't make it in the first hundred, the tool assumed there is no known licence template, and showed the relevant error message.
The quick fix would be probably increasing the number of templates fetched for a page (API's limit being 500) but I doubt it is a way to go. API offers a possibility to continue the previous query if number of stuffs to return is bigger than requested amount (100 in our case).
This patch intends to use this feature of the API. (At least) in theory, it should be used for all other API calls tool is doing, not only templates, but as I find it rather unlikely that it makes any sense in other types of queries, only template query uses it for now.

I am not super happy with the way those recursive deferred/promise objects are done here, so I'd be glad to see some comments and suggestions how to make it cleaner.

WMDE-Fisch commented:
+1 hmm works and looks sane... I have no better idea for now probably would have done something similar and would also be not fully satisfied.

addshore commented:
+1 from me!

It might be a nice idea to have some sort of HARD limit on the recursion so that if something off happens it doesn't consume all of the memory and make all of the requests!

wmde-manicki commented:
Good point, I'll add a limit to max 10 recursive calls.

wmde-manicki commented:
Added a limit as suggested (thanks!). The "demo" instance has been also updated, so this could be tested (ie. that adding the limit does not break what has been done before)

jakobw commented:
Works and looks good!
